content/blog/2022/2022-01-25-annotation-processing.md

```diff
@@ -0,0 +1,851 @@
+
+Normally when we run JavaDoc on a program the annotation @Inherited would not show up in the documentation. But when we use the @Documented annotation as below '@Inherited' would show up in the output too.
+
+```java
+@Documented
+@EmployeeType
+public class Employee { ...}
+```
+
+3. @Repeatable - Allows multiple repeating Custom Annotations on a method, class, or field. For eg. if you had the Annotation @Role defined on Employee it could be used as:
+
+```java
+@Repeatable
+public @interface Role { .. }
+
+@Role("Manager")
+@Role("Analyst")
+public class Employee { ...}
+```
+
+4. @Target - Specifies at which element the Annotation can be used, for eg in the above example if the annotation Role was defined only for TYPE, then it could only be applied to a Class.
+
+```java
+@Target({ElementType.TYPE})
+public @interface Role { .. }
+```
+
+The various self-explanatory element types are - ANNOTATION_TYPE, CONSTRUCTOR, FIELD, LOCAL_VARIABLE, METHOD, PACKAGE, PARAMETER, TYPE (Class).
+
+5. @Retention - Specifies when the annotation is discarded.
+
+​	  SOURCE - The annotation is used at compile-time and discarded at run-time.
+
+​	  CLASS - The annotation is stored in the class file at compile time and discarded at run time.
+
+​	  RUNTIME - The annotation is retained at run-time.
+
+If we needed an annotation to only provide error checking at compile-time like @Override we would use SOURCE. If we need an annotation to provide functionality at run-time such as @Test in Junit we would use RUNTIME.
+
+## Annotation Categories
+
+1. Marker Annotations - They do not contain any members or data. For eg. @Override or @Test.
+   
+   ```java
+   public Class Employee {
+   
+   @Override
+   public String getDeptName();
+   }
+   ```
+
+2. Single Value Annotations - They contain only one member and the parameter is the value of the member. For eg: Company(name="ABC")
+
+   ```java
+   @Company("XYZ")
+   public Class Employee {
+   
+       @Override
+       public String getDeptName();
+   }
+   ```
+
+3. Full Annotations - They consist of multiple name value pairs. For eg. Company(name="ABC", city="XYZ")
+
+   ```java
+   @Company(name="ABC", city="XYZ")
+   public Class Employee {
+   
+   @Override
+   public String getDeptName();
+   }
+   ```
+
+4. Type Annotations - They are declared with @Target and specify at which element the Annotation can be used, for eg. 
+
+   @Target({ElementType.METHOD, ElementType.TYPE}) would apply the 		annotation both at the Method and Class level.
+
+## Annotation Processing Overview
+
+Annotation processing is a built-in functionality provided by Java for scanning and processing annotations at compile time. Custom Processors can be developed based on AbstractProcessor which is the Java class providing the core functionality. They can be used to provide functionality such as automated code generation.
+
+## Annotation Processing Implementation - A real-world example
```

Comment:
The example used in this section is way too complicated to explain annotation processing. It's pages and pages of text and code that no reader will take the time to understand. Also, a multi-module Maven project is not really ideal as a learning exercise in annotation processing. The TOC shows me this structure for the section:

* Core Module
* Annotation Processor Module
* init()
* process()

none of the section headers tell me anything about annotation processing. Instead, try to make the section headers tell a story about annotation processing. Something like:

* Example Use Case
* Writing a Plain Annotation Processor
* Writing an Annotation Processor with JavaPoet



Instead, please find a very simple annotation processing example that doesn't take the focus away from annotation processing. Maybe a processor that throws an error when it finds a `@Deprecated` annotation so that it fails builds when someone has used the annotation. Or something that simply counts the number of certain annotations

Reply:
Thanks for your comments! This gives me a better idea about the structure we want and also a good idea about the level of complexity we desire. Once done, going forward I will be use this completed article as a reference for future ones.

When I picked this subject, the Article mentioned 'Writing an Annotation Processor' so I thought the main focus was writing a real-world complex Annotation Processor for mid-level to advanced users. The portion on Annotations was just added to provide a recap to advanced users, who don't deep dive into Annotations on a daily basis, on the entire Annotations topic. Now since I had added an Intro to Annotations too, I changed the title to 'Introduction to Annotations and Annotation Processing'. I think this was where the confusion about the level of complexity came about.

Reply:
Also, I am thinking I can split this into 2 articles now. One on 'Introduction to Annotations and Annotation Processing' and the other the original topic of 'Writing an Annotation Processor'. The first one can use the simple '@Deprecated' example you suggested. In the second one, I can just devote the whole article just to the custom Annotation Processor and dedicate the whole article to breaking down the code into smaller snippets and going into greater detail. Let me know what you think. Thanks.